controllers/topology/desired_state.go

```diff
@@ -173,14 +173,49 @@ func computeControlPlane(_ context.Context, s *scope.Scope, infrastructureMachin
 	}
 
 	// Sets the desired Kubernetes version for the control plane.
-	// TODO: improve this logic by adding support for version upgrade component by component
-	if err := contract.ControlPlane().Version().Set(controlPlane, s.Blueprint.Topology.Version); err != nil {
+	version, err := computeControlPlaneVersion(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compute version of control plane")
+	}
+	if err := contract.ControlPlane().Version().Set(controlPlane, version); err != nil {
 		return nil, errors.Wrap(err, "failed to set spec.version in the ControlPlane object")
 	}
 
 	return controlPlane, nil
 }
 
+// computeControlPlaneVersion returns the desired control plane version.
+// If creating a new control plane, returns the version defined in tolopogy.
+// If a control plane already exists:
+// - If the control plane or any of the machine deployments are upgrading, returns the version from the current control plane.
+// - If not, returns the version defined in topology.
```

Comment:
Can we move these comments inline in the function?

Reply:
With time, the logic below is going to change, if the comments are inlined we can make sure that it gets updated at the same time the function is updated

Reply:
Good idea. I will break up the comment and have them near the corresponding sections.